rpi-5.9.y - latest KMS patches from Maxime

6by9 opened the pull request:
3 patches that have been sent to dri-devel.

With these patches I've been switching dual HDMI monitors around between resolutions without any failures :-)

mripard committed:
drm/vc4: hvs: Pull the state of all the CRTCs prior to PV muxing

The vc4 display engine has a first controller called the HVS that will
perform the composition of the planes. That HVS has 3 FIFOs and can
therefore compose planes for up to three outputs. The timings part is
generated through a component called the Pixel Valve, and the BCM2711 has 6
of them.

Thus, the HVS has some bits to control which FIFO gets output to which
Pixel Valve. The current code supports that muxing by looking at all the
CRTCs in a new DRM atomic state in atomic_check, and given the set of
contraints that we have, assigns FIFOs to CRTCs or reject the mode
entirely. The actual muxing will occur during atomic_commit.

However, that doesn't work if only a fraction of the CRTCs' state is
updated in that state, since it will ignore the CRTCs that are kept running
unmodified, and will thus unassign its associated FIFO, and later disable
it.

In order to make the code work as expected, let's pull the CRTC state of
all the enabled CRTC in our atomic_check so that we can operate on all the
running CRTCs, no matter whether they are affected by the new state or not.

Cc: Hoegeun Kwon <hoegeun.kwon@samsung.com>
Fixes: 87ebcd42fb7b ("drm/vc4: crtc: Assign output to channel automatically")
Signed-off-by: Maxime Ripard <maxime@cerno.tech>

mripard committed:
drm/vc4: kms: Assign a FIFO to enabled CRTCs instead of active

The HVS has three FIFOs that can be assigned to a number of PixelValves
through a mux.

However, changing that FIFO requires that we disable and then enable the
pixelvalve, so we want to assign FIFOs to all the enabled CRTCs, and not
just the active ones.

Fixes: 87ebcd42fb7b ("drm/vc4: crtc: Assign output to channel automatically")
Signed-off-by: Maxime Ripard <maxime@cerno.tech>

6by9 commented:
I am sending review comments to the mailing list over "drm/vc4: crtc: Keep the previously assigned HVS FIFO".
As long as we're happy to revert and reapply, the patch (dropping it in the rebase), then I'm happy to merge ahead of the response. 

It's a malloc failure error path and code a coding style issue, so not likely to cause grief.

popcornmix commented:
On this tree don't worry about reverts and reapplies if a commit changes in review. We can drop them in a rebase.

6by9 commented on drivers/gpu/drm/vc4/vc4_crtc.c:
```diff
@@ -852,11 +852,18 @@ void vc4_crtc_destroy_state(struct drm_crtc *crtc,
+
+	vc4_crtc_state = kzalloc(sizeof(*vc4_crtc_state), GFP_KERNEL);
+	if (!vc4_crtc_state)
+		return;
```
Just FYI, my comment sent to the list on this is as follows:

This error path has me worried, but I may have missed it.

If crtc->state was set then it's been freed via
vc4_crtc_destroy_state. However I don't see anything that has reset
crtc->state to NULL.
If the new alloc fails then I believe we exit with crtc->state still
set to the old freed pointer.

The old code directly set crtc->state, so it got reset to NULL by the failure.

(There are also a couple of checkpatch complaints)

6by9 commented:
> On this tree don't worry about reverts and reapplies if a commit changes in review. We can drop them in a rebase.

I thought as much, but thought I ought to flag it.

mripard committed:
drm/vc4: crtc: Rework a bit the CRTC state code

The current CRTC state reset hook in vc4 allocates a vc4_crtc_state
structure as a drm_crtc_state, and relies on the fact that vc4_crtc_state
embeds drm_crtc_state as its first member, and therefore can be safely
casted.

However, this is pretty fragile especially since there's no check for this
in place, and we're going to need to access vc4_crtc_state member at reset
so this looks like a good occasion to make it more robust.

Signed-off-by: Maxime Ripard <maxime@cerno.tech>
Tested-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

mripard committed:
drm/vc4: crtc: Keep the previously assigned HVS FIFO

The HVS FIFOs are currently assigned each time we have an atomic_check
for all the enabled CRTCs.

However, if we are running multiple outputs in parallel and we happen to
disable the first (by index) CRTC, we end up changing the assigned FIFO
of the second CRTC without disabling and reenabling the pixelvalve which
ends up in a stall and eventually a VBLANK timeout.

In order to fix this, we can create a special value for our assigned
channel to mark it as disabled, and if our CRTC already had an assigned
channel in its previous state, we keep on using it.

Fixes: 87ebcd42fb7b ("drm/vc4: crtc: Assign output to channel automatically")
Signed-off-by: Maxime Ripard <maxime@cerno.tech>
Tested-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

mripard committed:
drm/atomic: Pass the full state to CRTC atomic enable/disable

If the CRTC driver ever needs to access the full DRM state, it can't do so
at atomic_enable / atomic_disable time since drm_atomic_helper_swap_state
will have cleared the pointer from the struct drm_crtc_state to the struct
drm_atomic_state before calling those hooks.

In order to allow that, let's pass the full DRM state to atomic_enable and
atomic_disable. The conversion was done using coccinelle, built tested on
all the drivers and actually tested on vc4.

Signed-off-by: Maxime Ripard <maxime@cerno.tech>

mripard committed:
drm/vc4: hvs: Align the HVS atomic hooks to the new API

Since the CRTC setup in vc4 is split between the PixelValves/TXP and the
HVS, only the PV/TXP atomic hooks were updated in the previous commits, but
it makes sense to update the HVS ones too.

Signed-off-by: Maxime Ripard <maxime@cerno.tech>

mripard committed:
drm/vc4: Pass the atomic state to encoder hooks

We'll need to access the connector state in our encoder setup, so let's
just pass the whole DRM state to our private encoder hooks.

Signed-off-by: Maxime Ripard <maxime@cerno.tech>

mripard committed:
drm/vc4: hdmi: Create a custom connector state

When run with a higher bpc than 8, the clock of the HDMI controller needs
to be adjusted. Let's create a connector state that will be used at
atomic_check and atomic_enable to compute and store the clock rate
associated to the state.

Signed-off-by: Maxime Ripard <maxime@cerno.tech>

mripard committed:
drm/vc4: hdmi: Store pixel frequency in the connector state

The pixel rate is for now quite simple to compute, but with more features
(30 and 36 bits output, YUV output, etc.) will depend on a bunch of
connectors properties.

Let's store the rate we have to run the pixel clock at in our custom
connector state, and compute it in atomic_check.

Signed-off-by: Maxime Ripard <maxime@cerno.tech>

6by9 committed:
drm/vc4: Fixup firmware-kms after "drm/atomic: Pass the full state to CRTC atomic enable/disable"

Prototype for those calls changed, so amend fkms (which isn't
upstream) to match.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

mripard committed:
drm/vc4: hdmi: Block odd horizontal timings

The FIFO between the pixelvalve and the HDMI controller runs at 2 pixels
per clock cycle, and cannot deal with odd timings.

Let's reject any mode with such timings.

Signed-off-by: Maxime Ripard <maxime@cerno.tech>

popcornmix commented:
Feel free to merge this if you are happy. I've been running with it locally.

6by9 commented:
Maxime's just pushed the changes for
-10 & 12 bit output 
- filtering 1366x768
- avoiding wifi frequencies.
I'm testing those and may as well add them to this.

pelwell commented:
OK - kick me when they're ready.

6by9 commented:
Extra commits pushed to this branch.
I'm rebuilding to test, although I don't believe I have any displays that can take 10 or 12bit input.

HiassofT commented on drivers/gpu/drm/vc4/vc4_hdmi.c:
```diff
@@ -1128,6 +1189,11 @@ static int vc4_hdmi_encoder_atomic_check(struct drm_encoder *encoder,
 		pixel_rate *= 2;
 
+	if (conn_state->max_bpc == 12)
+		pixel_rate *= 150 / 100;
```
this looks odd - doesn't 150/100 just evaluate to 1, so that line is a no-op?

same thing 2 lines below

6by9 replied:
Quite right.
I've just pulled them in off the dri-devel mailing list - they haven't been reviewed yet. Feel free to comment in your own right.

popcornmix replied:
Yep...

6by9 replied:
I'm emailing Maxime with a heads up on this one.

HiassofT replied:
thanks!

pelwell commented:
This PR is just waiting on a fix for the very rounded `*= a/b` scaling.

6by9 commented:
> This PR is just waiting on a fix for the very rounded *= a/b scaling.

Yes, I was giving it a day to see if Maxime came back over it. There's also been a query over the wifi patch on dri-devel.

If I don't hear anything by close of play then I'll drop the 10/12bit and wifi patches and run with that for now.

stevenhoving commented:
@pelwell is it really the case that this PR is just waiting on a fix for something as trivial as this?

```
if (conn_state->max_bpc == 12)
	pixel_rate = (pixel_rate * 150) / 100;
else if (conn_state->max_bpc == 10)
	pixel_rate = (pixel_rate * 125) / 100;
```

Or this PR is waiting for a upstream fix for this problem; so waiting for Maxime to publish the upstream fix?

popcornmix commented:
Yes, the fix is trivial, but ideally we'd like to get the same fix as is merged upstream so we don't get conflicts when we rebase.

There's also the slight concern that any testing done didn't include the pixel clock boost so that should really be double checked.

stevenhoving commented:
> Yes, the fix is trivial, but ideally we'd like to get the same fix as is merged upstream so we don't get conflicts when we rebase.
> 
> There's also the slight concern that any testing done didn't include the pixel clock boost so that should really be double checked.

Thanks for clarifying

6by9 commented:
PR updated to drop those patches. Hopefully the others won't change.

pelwell commented:
I'll take that chance.
